cellxgene_schema_cli/cellxgene_schema/matrix_utils.py

```diff
@@ -0,0 +1,135 @@
+    ]
+
+    would return a NumPy array of [3, 12, 1]
+    """
+    if isinstance(matrix, np.ndarray):
+        return matrix.sum(axis=0).ravel()
+
+    if sparse.issparse(matrix):
+        return np.array(matrix.sum(axis=0)).ravel()
+
+    # Handle Dask array (could be sparse or dense)
+    if isinstance(matrix, DaskArray):
+
+        def sum_columns(chunk):
+            if sparse.issparse(chunk):
+                return np.array(chunk.sum(axis=0)).ravel()
+            else:
+                return chunk.sum(axis=0)
+
+        # If chunks along axis=0 > 1, we need to sum per block then aggregate
+        if len(matrix.chunks[0]) > 1:
+            partial_sums = map_blocks(sum_columns, matrix, drop_axis=0, dtype=float)
+            return partial_sums.sum(axis=0).compute().ravel()
+        else:
+            return sum_columns(matrix.compute())
+
+    raise TypeError(f"Unsupported matrix type: {type(matrix)}")
+
+
+def count_matrix_nonzero(matrix: DaskArray) -> int:
+    def count_nonzeros(matrix_chunk: Union[np.ndarray, sparse.spmatrix], is_sparse_matrix: bool) -> np.array:
+        nnz = matrix_chunk.nnz if is_sparse_matrix else np.count_nonzero(matrix_chunk)
+        return np.array([nnz])
+
+    is_sparse_matrix = get_matrix_format(matrix) in SPARSE_MATRIX_TYPES
+    if len(matrix.chunks[0]) > 1:
+        nonzeros = map_blocks(count_nonzeros, matrix, is_sparse_matrix, drop_axis=1, dtype=int).compute().sum()
+    else:
+        nonzeros = count_nonzeros(matrix.compute(), is_sparse_matrix)[0]
+    return nonzeros
+
+
+def check_non_csr_matrices(adata: ad.AnnData):
+    """
+    Check X, raw.X and layers matrices for having more than 50% zeros and not being csr_matrix
+
+    If found, convert to csr_matrix
+    """
+
+    def get_sparsity(matrix: DaskArray, format: str):
+        is_sparse_matrix = format in SPARSE_MATRIX_TYPES
+        nnz = count_matrix_nonzero(matrix, is_sparse_matrix)
+        sparsity = 1 - nnz / np.prod(matrix.shape)
+        return sparsity
+
+    format = get_matrix_format(adata.X)
+    if format != "csr" and get_sparsity(adata.X, format) >= 0.5:
+        adata.X = adata.X.map_blocks(sparse.csr_matrix, dtype=adata.X.dtype)
+
+    if adata.raw is not None:
+        format = get_matrix_format(adata.raw.X)
+        if format != "csr" and get_sparsity(adata.raw.X, format) >= 0.5:
+            raw_adata = ad.AnnData(adata.raw.X, var=adata.raw.var, obs=adata.obs)
+            raw_adata.X = raw_adata.X.map_blocks(sparse.csr_matrix, dtype=raw_adata.X.dtype)
+            adata.raw = raw_adata
+            del raw_adata
+
+    for layer in adata.layers:
+        format = get_matrix_format(adata.layers[layer])
+        if format != "csr" and get_sparsity(adata.layers[layer], format) >= 0.5:
+            adata.layers[layer] = adata.layers[layer].map_blocks(sparse.csr_matrix, dtype=adata.layers[layer].X.dtype)
+
+    return adata
+
+
+def debug_print_matrix(matrix: ad.AnnData, matrix_name: str, max_rows=20, max_cols=20):
```

Comment:
i added this to debug things earlier, not actually using it in actual code i'm committing. but i figured it could be useful in the future too